### ISSUE TITLE
Option to disable inheritance

### ISSUE THREAD

**Issue by kriko** [gitlabform/gitlabform#842]
GitLabForm has a feature to apply settings from a parent group to sub-groups, however this can lead to unexpected configuration that is not desired.

We are using GitLabForm to manage configuration of a hierarchical GitLab setup, with many levels of sub-groups. However, when a parent group as a set of say, badges, these will also be configured for sub-groups - this in turn ends up with duplicate or triple amount of badges for GitLab projects (since badges are additive in GitLab by default - parent group badges are displayed on child levels).

One way to work around this issue is by adding `inherit: false` option to the sub-group level, however this quickly becomes a nuisance, because it would require adding this configuration parameter to all sub-groups and if we add it for a configuration element that is not present in the parent, we quickly get the following error:
```
Error: The inheritance-break flag set in "parent/sub1/sub2/sub3" under key "group_members" is invalid
because it has no higher level setting to inherit from.
```

Thus, to work around this issue, I propose a global level configuration for GitLabForm, to disable inheritance for all properties.

eg.
```
config_version: 3
inherit: false # Disable config inheritance on a global level

gitlab:
  url: https://gitlab.yourcompany.com
  # alternatively use the GITLAB_TOKEN environment variable
  token: "<private token of an admin user>"

```

**Comment by amimas** [comment]
Hi @kriko - Thanks for opening the issue.

> GitLabForm has a feature to apply settings from a parent group to sub-groups, however this can lead to unexpected configuration that is not desired.

I want to make a correction here because the terminologies being used can be confusing or misleading to someone who's new (or not) to gitlab or gitlabform.

GitLabForm does not have a feature to apply settings from "parent group" to "sub-group". That's a feature of GitLab itself. You configure a particular setting at the group level (i.e. branch protection setting) and that setting gets applied to (or inherited by) all the projects under that group, including those in the subgroup.

GitLabForm applies settings to a group or project as per the config file used for its execution. To minimize/simply the config, it has a feature of writing configs using a hierarchy. When it runs, it basically generates the effective config from that hierarchy for every entity and then applies the settings accordingly. For more details, see the [config reference here](https://gitlabform.github.io/gitlabform/reference/#projects-and-groups-configuration).

As for your suggestion of adding a feature to config syntax, let's revisit the solution later. You gave an example of an issue when configuring badges. Could you please share a sample config to clarify the issue? I'm wondering whether it's a feature for `badges` config as opposed to a global option, which can introduce other unintended consequences.

**Comment by rrotaru** [comment]
I just started using Gitlabform and I wanted to add some context here on what I believe @kriko was calling out:

Currently, group config can only be specified via the `/*` path ending ([code](https://github.com/gitlabform/gitlabform/blob/2d7589121be6b8677f2adef2e3822b877698276c/gitlabform/configuration/groups.py#L18-L29)). This means that with a configuration such as this:

```yml
projects_and_groups:
  "group-a/subgroup-b/*":
    saml_group_links:
      enforce: true
      group_label:
        saml_group_name: my_saml_group
        access_level: developer
```

Running `gitlabform ALL` would _also_ *directly* apply the SAML group link configuration to any sub-subgroups `group-a/subgroup-b/subsubgroup-c`. This is not a default behavior of GitLab itself, and to avoid configuring nested SAML group links, the config would need to define and apply `inherit: false` for every subgroup.

Right now `ALL` and `ALL_DEFINED` terminology is a bit weird, because running `gitlabform ALL_DEFINED` on the above config would _not_ recurse the changes down to `group-a/subgroup-b/subsubgroup-c` even though there is a wildcard `/*` ending in that config.

I think an option to apply group settings to a group defined in the hierarchy _without_ recursing to subgroups would solve this.

**Comment by ShepleySound** [comment]
Just wanted to add onto this issue, because I believe what I'm seeing is closely related, and hopefully the examples clear up some confusion around this behavior. **Yes**, GitLab passes configuration down to subgroups, but GitLabForm is _also_ explicitly adding configuration to subgroups depending on the context in a way that I'd argue is a little unintuitive.

I'm using this command, for reference:

```shell
gitlabform -c config.yml ALL_DEFINED --diff-only-changed --noop --output-file "config-final.yml"
```

## Config 1

```yml
config_version: 3

gitlab:
  url: https://gitlab.example.com

.webhook: &webhook
  group_hooks:
    "https://example.com/webhook":
      merge_request_events: true

.dev_access: &dev_access
  members:
    users:
      dev-1:
        access_level: developer

projects_and_groups:
  group-a/subgroup-a/*:
    <<: *webhook
  group-a/subgroup-a/sub-subgroup-a/*:
    <<: *dev_access
```

## Config 2

```yml
# Only difference is that we've removed the `sub-subgroup-a` block
config_version: 3

gitlab:
  url: https://gitlab.example.com

.webhook: &webhook
  group_hooks:
    "https://example.com/webhook":
      merge_request_events: true

.dev_access: &dev_access
  members:
    users:
      dev-1:
        access_level: developer

projects_and_groups:
  group-a/subgroup-a/*:
    <<: *webhook
```

## Config 1 Final Output

```yml
%YAML 1.2
---
group-a/subgroup-a:
   group_hooks:
      https://example.com/webhook:
         merge_request_events: true
group-a/subgroup-a/sub-subgroup-a:
   group_hooks:
      https://example.com/webhook:
         merge_request_events: true
group-a/subgroup-a/sub-subgroup-a/project-a:
   members:
      users:
         dev-1:
            access_level: 30
group-a/subgroup-a/sub-subgroup-a/project-b:
   members:
      users:
         dev-1:
            access_level: 30
```

## Config 2 Final Output
```yml
%YAML 1.2
---
group-a/subgroup-a:
   group_hooks:
      https://example.com/webhook:
         merge_request_events: true
```

So it seems like `ALL_DEFINED` is pretty literal here - Because we included `group-a/subgroup-a` in the configuration, it's also inheriting the group webhook configuration, even though it looks like, at first glance, we only meant to configure it in the parent group. So we end up with both groups having the webhook explicitly configured, which doubles the surface area for drift to happen.

Where I think it gets confusing is when we leave out any mention of `sub-subgroup-a`. Suddenly there's no recursion. I could see how in larger configurations this could lead to a lot of confusion, because the recursion changes depending on whether the child is present in the config, but not based on the configuration of the child itself.

Further, in my example there doesn't seem to be a way to break inheritance.

```yml
.webhook: &webhook
  group_hooks:
    inherit: false
    "https://example.com/webhook":
      merge_request_events: true
```

Results in

```log
Error: The inheritance-break flag set in "group-a/subgroup-a" under key "group_hooks" is invalid because it has no higher level setting to inherit from.
```

**Comment by amimas** [comment]
Hey everyone,

Thanks for your input to clarify the issue. So, I can definitely see unwanted behaviour using config like below:

```yaml
projects_and_groups:
  group-a/*:
    saml_group_links:
      enforce: true
      group_label:
        saml_group_name: my_saml_group
        access_level: developer
    group_hooks:
      "https://example.com/webhook":
       merge_request_events: true
```

With the above config saml group linking and group webhook will be applied to the top level group `group-a`. Issue here is that if there are other subgroups (i.e. `group-a/subgroup-1/`), that subgroup or others will also have the same config. That is not desired.

GitLabForm already allows breaking inheritance via `inherit: false` flag. So, the config should probably look like this:

```diff
projects_and_groups:
  group-a/*:
    saml_group_links:
      enforce: true
      group_label:
        saml_group_name: my_saml_group
        access_level: developer
    group_hooks:
      "https://example.com/webhook":
       merge_request_events: true
+ group-a/subgroup-1/*:
+   saml_group_links:
+     inherit: false
+   group_hooks:
+     inherit: false
```

Using above config, now saml group linking and webhook will only be applied to `group-a`. It won't be applied to `group-a/subgroup-1` or even `group-a/subgroup-1/subsubgroup-1`.

Yes, it requires few extra lines of config?

> One way to work around this issue is by adding inherit: false option to the sub-group level, however this quickly becomes a nuisance, because it would require adding this configuration parameter to all sub-groups and if we add it for a configuration element that is not present in the parent, we quickly get the following error:

I'm not sure that's true. Using my example above, are you saying the inheritance breaking feature doesn't work for `group-a/subgroup-1/subsubgroup-1`? Can you please confirm?

**Comment by ShepleySound** [comment]
A few extra lines of config turns into _a lot_ when managing configs for hundreds (or thousands) of groups.

This also results in two fairly hard-to-miss bugs, both related to LDAP Group Links -

## Duplicate Error

### Config

```yaml
projects_and_groups:
  test/a/*:
    group_ldap_links:
      my_group_is_developer:
        provider: ldapmain
        cn: my-group
        group_access: developer
  test/a/b/*:
    group_ldap_links:
      my_group_is_owner:
        provider: ldapmain
        cn: my-group
        group_access: owner
```

### Rendered output

```yaml
%YAML 1.2
---
test/a:
   group_ldap_links:
      my_group_is_developer:
         provider: ldapmain
         cn: my-group
         group_access: 30
test/a/b:
   group_ldap_links:
      my_group_is_developer:
         provider: ldapmain
         cn: my-group
         group_access: 30
      my_group_is_owner:
         provider: ldapmain
         cn: my-group
         group_access: 50
```

The above will render with `--noop -o`, but will fail if you try to apply it with the following error:

> Error: Entities my_group_is_developer and my_group_is_owner in group_ldap_links for test/a/b are the same in terms of their defining keys: ('provider' and ('cn' or 'filter'))

So you **must** use `inherit: false` with `group_ldap_links`, otherwise this error might occur in cases where you're trying to "stack" LDAP groups with different permissions. But the logic of the original config is sound - we want my-group to be a developer on a parent group, but an owner on a specific sub-group.

But in a complex configuration, someone might call out `test/a/b/*` to add some configuration completely unrelated to `group_ldap_links` and not know that they need to care about inheritance, which leads to the next problem...

## LDAP breaks direct membership

GitLab, by design, does not allow direct memberships and LDAP group sync to coexist in the same GitLab Group. If you enable group sync, then anyone who is a direct member of the group is removed. This is [called out in the docs as a warning](https://docs.gitlab.com/user/group/access_and_permissions/#manage-group-memberships-with-ldap). But just because you add `group_ldap_links` to `test/a` does **not** mean you want it to be used in `test/a/b`. Making that assumption has some pretty nasty consequences.

So if you want to **safely** manage `group_ldap_links` with GitLabForm, you have to basically enforce that EVERY entry in your configuration ending with `/*` must do

```
test/a/b/*:
  ...
  group_ldap_links:
    inherit: false
```

I think that this behavior should at least be called out in the documentation, since a missing flag could lead to permission structures being destroyed. But given that there's almost no reason you would ever want these group configurations to inherit down the tree, `inherit: false` really should be the default behavior for them...

**Comment by itewk** [comment]
having same issue, with badges. there is no way currently to configure a group_badge only at the top level group. (let alone any other setting). 

> Yes, it requires few extra lines of config? (context: adding `inherit: false` to every sub group)

As previously stated by others, this doesn't work when you have 100s / of various nested groups that you don't want to explicitly track because you just want to rely on the wildcard syntax for that part.

it would be nice if there was a difference between:

```
  # ==============================================================================
  # Group settings for this group only
  # ==============================================================================
  administration:
    group_badges:
      pipeline-status:
        name: "Pipeline Status"
        link_url: "https://gitlab.example.xyz/%{project_path}/"
        image_url: "https://gitlab.example.xyz/%{project_path}/badges/%{default_branch}/pipeline.svg"
      enforce: true
```

and

```
  # ==============================================================================
  # Group settings for this group and all nested groups and projects
  # ==============================================================================
  administration/*:
    bla: ...
```

basically the suggestion is that specifying not `/*` works the same for groups as it does for projects, only applys against that specific resource.

thoughts?

**Comment by itewk** [comment]
i can get pretty close to what i want by passing `ALL_DEFINED` and `--recurse-subgroups` the only problem is IF i do put in explicate config for a sub-group then it will inherit the parent group config, in which case i have to put in the inherit false or skip true. but thats manageable for me because its not an often case will be explicitly specifying a sub-group

**Comment by amimas** [comment]
Thanks @itewk for opening the PR #1429. I agree this feature is needed.

I think it is a decent amount of work and requires some thoughts, and hoping this can be addressed properly with everyone's help. This will resolve multiple issues.

I wanted to get your thoughts on the implementation details. In your PR (and the one it based on) is suggesting config syntax like below, where path is used to determine where config will apply:

```yaml
    projects_and_groups:
      "*":    # Applies to all groups, subgroups, projects
        group_settings:
          setting_from_common: common_value
        project_settings:
          ....

      some_group/*:  # Applies to all groups, subgroups, and projects under `some_group/*` path
        group_settings:
          visibility: private
          description: "recursive description"
        project_settings:
          ....

      some_group/:   # Applies to `some_group` GROUP ONLY - won't support configs for projects under it and subgroups
        group_settings:
          visibility: public
          ai_settings_attributes:
            ai_catalog_restricted_to_group_hierarchy: true

      some_group/subgroup/:  # Applies to `some_group/subgroup` GROUP ONLY - won't support configs for projects under it and subgroups
        group_settings:
          description: "subgroup only description"
```

Please let me know if I misunderstood something.

PR #1301 is trying to deal with inheritance and has a different proposal. This one, I believe, is introducing a new config key called `propagate: false` - similar to `inherit: false`. You'll set it at the top level or where needed and that "section" of the config won't be included into downstream paths (i.e. subgroups or projects).

I think both approach are interesting. Would love to get feedbacks on which approach would be best. Clearly we need a solution here and I'm not sure it's trivial. Appreciate everyone's patience and help with this.

cc: @TimKnight01 @rickbrouwer

**Comment by itewk** [comment]
@amimas your description of #1429 is correct.

as for propgate and inherit attributes, those don't meet my neads because i dont want to exclude entire sections from propagating, only specific attributes, and unless there started to be nested attributes for each attribute, that doesn't work.

the pr  #1429 allows to specify any number of group attributes that apply to just a specific group, and/or the current functionality of any number of attributes that apply to all the nested groups.

i am running #1429 fork in my prod environment and it is working great for dealing with the scenarios where it is useful. such as https://github.com/gitlabform/gitlabform/issues/1427

**Comment by amimas** [comment]
Thanks @itewk for looking into it.

>  i dont want to exclude entire sections from propagating, only specific attributes

I'm not actually following that. Maybe I missed something but I don't see how the following config is for specific **attributes** only - not *section*:

```yaml
      some_group/*:  # Applies to all groups, subgroups, and projects under `some_group/*` path
        group_settings:
          visibility: private
          description: "recursive description"
          ai_settings_attributes:
            ai_catalog_restricted_to_group_hierarchy: true

      some_group/:   # Applies to `some_group` GROUP ONLY - won't support configs for projects under it and subgroups
        group_settings:
          visibility: public
          description: "hello world"
          ai_settings_attributes:
            ai_catalog_restricted_to_group_hierarchy: true
```

In the above example, `some_group` setting will be overwritten for everything in the config (i.e. `visibility`, `description`, `ai_settings_attributes`. Isn't it?

**Comment by itewk** [comment]
@amimas apologies for the confusion. When I said "i dont want to exclude entire sections from propagating, only specific attributes" i meant that doing the `inherit` or `propgate` wont work because you can only apply those to a section, so like you can so don't inherit or propagate any of the group_settings, but there is no way to use those attributes to say, don't propagate/inherit this one specific setting.

with the approach proposed here you can do that because you can specify the individual attributes either for the specific group (`/`) or for propagating to all sub groups (`/*`).
